File: src/docker-compose-runner.ts
import { promisify } from "util";
import { exec } from "child_process";
import * as path from "path";
import { createDockerComposeEnv } from "./docker-config.js";

const execPromise = promisify(exec);

/**
 * Render R Markdown using docker-compose
 * 
 * @param rmdDir Directory containing R Markdown files
 * @param filename Filename of the R Markdown file to render
 * @param outputFormat Output format (html, pdf, or word)
 * @returns Output filename
 */
export async function renderWithDockerCompose(
  rmdDir: string,
  filename: string,
  outputFormat: "html" | "pdf" | "word" = "html"
): Promise<string> {
  const inputFile = path.join(rmdDir, filename);
  
  // Create environment variables for docker-compose
  const env = createDockerComposeEnv(rmdDir, filename, outputFormat);
  
  // Add the render command based on output format
  const renderCommand = outputFormat === "html" 
    ? "render" 
    : outputFormat === "pdf" 
      ? "render_pdf" 
      : "render_word";
  
  env.RENDER_COMMAND = renderCommand;
  
  // Build the environment variables string for the command
  const envString = Object.entries(env)
    .map(([key, value]) => `${key}=${value}`)
    .join(" ");
  
  try {
    // Run docker-compose with the environment variables
    const { stdout, stderr } = await execPromise(`${envString} docker-compose up --build`);
    
    // Extract the output file path from logs
    const outputFileMatch = stdout.match(/Rendered file: ([^\n]+)/);
    if (!outputFileMatch) {
      throw new Error("Failed to extract output file path from logs");
    }
    
    const outputFilePath = outputFileMatch[1];
    const outputFileName = path.basename(outputFilePath);
    
    return outputFileName;
  } catch (error: any) {
    console.error("Error rendering with docker-compose:", error);
    throw new Error(`Failed to render R Markdown file with docker-compose: ${error.message || String(error)}`);
  } finally {
    // Clean up by stopping and removing containers
    try {
      await execPromise("docker-compose down");
    } catch (cleanupError) {
      console.error("Error cleaning up docker-compose:", cleanupError);
    }
  }
}

File: docker_compose_runner.go
package main

import (
	"fmt"
	"os"
	"os/exec"
	"path/filepath"
	"regexp"
)

// RenderWithDockerCompose renders R Markdown using docker-compose
func RenderWithDockerCompose(rmdDir, filename, outputFormat string) (string, error) {
	// We don't need to use inputFile directly as it's handled by the Docker environment
	_ = filepath.Join(rmdDir, filename)

	// Create environment variables for docker-compose
	env := CreateDockerComposeEnv(rmdDir, filename, outputFormat)

	// Add the render command based on output format
	renderCommand := "render"
	if outputFormat == "pdf" {
		renderCommand = "render_pdf"
	} else if outputFormat == "word" {
		renderCommand = "render_word"
	}

	env["RENDER_COMMAND"] = renderCommand

	// Build the environment variables for the command
	cmd := exec.Command("docker-compose", "up", "--build")
	cmd.Env = os.Environ()
	for key, value := range env {
		cmd.Env = append(cmd.Env, fmt.Sprintf("%s=%s", key, value))
	}

	// Capture stdout and stderr
	output, err := cmd.CombinedOutput()
	if err != nil {
		return "", fmt.Errorf("failed to run docker-compose: %w\nOutput: %s", err, string(output))
	}

	// Extract the output file path from logs
	outputStr := string(output)
	re := regexp.MustCompile(`Rendered file: ([^\n]+)`)
	matches := re.FindStringSubmatch(outputStr)
	if len(matches) < 2 {
		return "", fmt.Errorf("failed to extract output file path from logs")
	}

	outputFilePath := matches[1]
	outputFileName := filepath.Base(outputFilePath)

	// Clean up by stopping and removing containers
	cleanupCmd := exec.Command("docker-compose", "down")
	if err := cleanupCmd.Run(); err != nil {
		fmt.Printf("Warning: failed to clean up docker-compose: %v\n", err)
	}

	return outputFileName, nil
}
